typing/ctype.ml

```diff
@@ -1305,67 +1300,56 @@ let instance_class params cty =
 
 (**** Instantiation for types with free universal variables ****)
 
-let rec diff_list l1 l2 =
-  if l1 == l2 then [] else
-  match l1 with [] -> invalid_arg "Ctype.diff_list"
-  | a :: l1 -> a :: diff_list l1 l2
-
-let conflicts free bound =
-  let bound = List.map get_id bound in
-  TypeSet.exists (fun t -> List.memq (get_id t) bound) free
-
 let delayed_copy = ref []
     (* copying to do later *)
+let add_delayed_copy t copy_scope ty =
+  delayed_copy :=
+    lazy (Transient_expr.set_stub_desc t (Tlink (copy copy_scope ty))) ::
+    !delayed_copy
 
 (* Copy without sharing until there are no free univars left *)
 (* all free univars must be included in [visited]            *)
-let rec copy_sep ~copy_scope ~fixed ~free ~bound ~may_share
-    (visited : (int * (type_expr * type_expr list)) list) (ty : type_expr) =
+let rec copy_sep ~copy_scope ~fixed ~free ~may_share
+    ~(visited : type_expr TypeHash.t) (ty : type_expr) =
   let univars = free ty in
   if is_Tvar ty || may_share && TypeSet.is_empty univars then
     if get_level ty <> generic_level then ty else
     let t = newstub ~scope:(get_scope ty) in
-    delayed_copy :=
-      lazy (Transient_expr.set_stub_desc t (Tlink (copy copy_scope ty)))
-      :: !delayed_copy;
+    add_delayed_copy t copy_scope ty;
     t
   else try
-    let t, bound_t = List.assq (get_id ty) visited in
-    let dl = if is_Tunivar ty then [] else diff_list bound bound_t in
-    if dl <> [] && conflicts univars dl then raise Not_found;
-    t
+    TypeHash.find visited ty
   with Not_found -> begin
     let t = newstub ~scope:(get_scope ty) in
     let desc = get_desc ty in
-    let visited =
-      match desc with
-        Tarrow _ | Ttuple _ | Tvariant _ | Tconstr _ | Tobject _ | Tpackage _ ->
-          (get_id ty, (t, bound)) :: visited
-      | Tvar _ | Tfield _ | Tnil | Tpoly _ | Tunivar _ ->
-          visited
-      | Tlink _ | Tsubst _ ->
-          assert false
-    in
-    let copy_rec = copy_sep ~copy_scope ~fixed ~free ~bound visited in
+    begin match desc with
+      Tarrow _ | Ttuple _ | Tvariant _ | Tconstr _ | Tobject _ | Tpackage _ ->
+        TypeHash.add visited ty t
+    | Tvar _ | Tfield _ | Tnil | Tpoly _ | Tunivar _ ->
+        ()
```

Comment:
why are all these types excluded from being added to `visited`?

Reply:
This was a left-over of the previous code; we didn't want to create recursion in a strange place. Now we are sharing everything, there is no point of doing this selective cacheing, so I removed it, and just add everything.
By the way, the invariants guarantee that `Tvar`, `Tunivar` and `Tnil` cannot occur here anyway. And I believe that `Tfield` and `Tpoly` cannot be shared without sharing there parent, so this shouldn't make a difference.